File: pharmacy-api/services/goods_service/controller/controller.go
package controller

import (
	"log"
	"math/rand"
	models "pharmacy-api/shared/models"
	"strconv"
	"time"

	"gorm.io/gorm"
)

func GetGoods(db *gorm.DB, query string, pageStr string, limitStr string) (result map[string]any, err error) {
	limit, strErr := strconv.Atoi(limitStr)

	if strErr != nil || limit < 1 {
		limit = 10
	} else if limit > 40 {
		limit = 40
	}
	page, strErr := strconv.Atoi(pageStr)
	if strErr != nil || page < 1 {
		page = 1
	}
	offset := (page - 1) * limit
	var goodsByTags, goodsByProducers, goodsByName []models.Goods
	var totalCount int64
	var goods []models.Goods
	var seenIDs = make(map[uint]bool)
	if query != "" {
		db.Preload("Producer").
			Joins("JOIN goods_tags ON goods.id = goods_tags.goods_id").
			Joins("JOIN tags ON goods_tags.tag_id = tags.id").
			Where("tags.tag_name LIKE ?", "%"+query+"%").
			Order("name ASC").Find(&goodsByTags)

		db.Preload("Producer").
			Joins("JOIN producers ON goods.producer_id = producers.id").
			Where("producers.producer_name LIKE ?", "%"+query+"%").
			Order("name ASC").Find(&goodsByProducers)

		db.Preload("Producer").
			Where("name LIKE ?", "%"+query+"%").
			Order("name ASC").Find(&goodsByName)
		for _, item := range goodsByTags {
			if !seenIDs[item.ID] {
				goods = append(goods, item)
				seenIDs[item.ID] = true
			}
		}
		for _, item := range goodsByProducers {
			if !seenIDs[item.ID] {
				goods = append(goods, item)
				seenIDs[item.ID] = true
			}
		}
		for _, item := range goodsByName {
			if !seenIDs[item.ID] {
				goods = append(goods, item)
				seenIDs[item.ID] = true
			}
		}
		totalCount = int64(len(goods))
		goods = goods[offset:min(len(goods), offset+limit)]
	} else {
		db.Model(&goods).Count(&totalCount)
		db.Order("name ASC").Limit(limit).Offset(offset).Find(&goods)
	}
	totalPages := (totalCount + int64(limit) - 1) / int64(limit)

	var response []models.GoodsResponse
	for _, goodsitem := range goods {
		response = append(response, models.GoodsResponse{
			ID:          goodsitem.ID,
			Name:        goodsitem.Name,
			Image:       goodsitem.Image,
			IsInStock:   goodsitem.IsInStock,
			Description: goodsitem.Description,
			Price:       goodsitem.Price,
		})
	}

	result = make(map[string]any)
	result["Items"] = response
	result["TotalPages"] = totalPages
	result["CurrentPage"] = page

	return
}

func GetGoodsByID(db *gorm.DB, id int) (result map[string]any, err error) {

	var good models.Goods

	if err = db.Preload("Producer").Preload("Tags").Find(&good, id).Error; err != nil {
		return
	}

	var tagNames []string
	for _, tag := range good.Tags {
		tagNames = append(tagNames, tag.TagName)
	}

	response := models.GoodsResponse{
		ID:                   good.ID,
		Name:                 good.Name,
		Image:                good.Image,
		Producer:             good.Producer.ProducerName,
		IsInStock:            good.IsInStock,
		Tags:                 tagNames,
		Instruction:          good.Instruction,
		Description:          good.Description,
		IsPrescriptionNeeded: good.IsPrescriptionNeeded,
		Price:                good.Price,
	}

	result = make(map[string]any)
	result["Response"] = response

	return
}

func GetPromoItems(db *gorm.DB) (result map[string]any, err error) {
	var goods []models.Goods
	if err = db.Where("is_in_stock LIKE ?", "1").Find(&goods).Error; err != nil {
		return
	}
	totalGoods := len(goods)
	if totalGoods == 0 {
		return
	}
	rand.NewSource(time.Now().UnixNano())
	indices := rand.Perm(totalGoods)[:min(5, totalGoods)]
	var promoItems []models.PromoItem
	for _, index := range indices {
		promoItems = append(promoItems, models.PromoItem{
			ID:          goods[index].ID,
			Name:        goods[index].Name,
			Description: goods[index].Description,
			Price:       goods[index].Price,
			Image:       goods[index].Image,
		})
	}

	result = make(map[string]any)
	result["Response"] = promoItems

	return
}

func UpdateGoods(db *gorm.DB, id int, updateData models.GoodsUpdateRequest, claims models.Claims) (result map[string]any, err error) {

	var existingGood models.Goods
	if err = db.First(&existingGood, id).Error; err != nil {
		log.Println("Good error [" + claims.Username + "]" + err.Error())
		return
	}

	existingGood.Name = updateData.Name
	existingGood.Instruction = updateData.Instruction
	existingGood.Description = updateData.Description
	existingGood.IsPrescriptionNeeded = updateData.IsPrescriptionNeeded
	existingGood.IsInStock = updateData.IsInStock
	existingGood.Price = updateData.Price

	if err = db.Save(&existingGood).Error; err != nil {
		log.Println("Good error [" + claims.Username + "]" + err.Error())
		return
	}

	log.Println("Good PATH [" + claims.Username + "]")

	result = make(map[string]any)
	result["Response"] = "Good updated"
	return
}

File: pharmacy-api/src/services/goods_service/controller/controller.ts
import { DataSource } from 'typeorm';
import { Goods } from '@pharmacy/src/shared/models/database_models';
import { GoodsResponse, PromoItem, GoodsUpdateRequest } from '@pharmacy/src/shared/models/responses';
import { Claims } from "@pharmacy/src/shared/models/models"

export async function getGoods(
  dataSource: DataSource,
  query: string,
  pageStr: string,
  limitStr: string
): Promise<{ Items: GoodsResponse[]; TotalPages: number; CurrentPage: number }> {
  let limit = parseInt(limitStr);
  if (isNaN(limit) || limit < 1) {
    limit = 10;
  } else if (limit > 40) {
    limit = 40;
  }

  let page = parseInt(pageStr);
  if (isNaN(page) || page < 1) {
    page = 1;
  }
  const offset = (page - 1) * limit;

  const goodsRepository = dataSource.getRepository(Goods);
  let goods: Goods[] = [];
  let totalCount = 0;

  if (query) {
    const goodsByTags = await goodsRepository
      .createQueryBuilder('goods')
      .leftJoinAndSelect('goods.producer', 'producer')
      .leftJoin('goods.goodsTags', 'goodsTags')
      .leftJoin('goodsTags.tag', 'tag')
      .where('tag.tagName LIKE :query', { query: `%${query}%` })
      .orderBy('goods.name', 'ASC')
      .getMany();

    const goodsByProducers = await goodsRepository
      .createQueryBuilder('goods')
      .leftJoinAndSelect('goods.producer', 'producer')
      .where('producer.producerName LIKE :query', { query: `%${query}%` })
      .orderBy('goods.name', 'ASC')
      .getMany();

    const goodsByName = await goodsRepository
      .createQueryBuilder('goods')
      .leftJoinAndSelect('goods.producer', 'producer')
      .where('goods.name LIKE :query', { query: `%${query}%` })
      .orderBy('goods.name', 'ASC')
      .getMany();

    const seenIDs = new Set<number>();
    const allGoods = [...goodsByTags, ...goodsByProducers, ...goodsByName];
    
    for (const item of allGoods) {
      if (!seenIDs.has(item.id)) {
        goods.push(item);
        seenIDs.add(item.id);
      }
    }

    totalCount = goods.length;
    goods = goods.slice(offset, offset + limit);
  } else {
    [goods, totalCount] = await goodsRepository
      .createQueryBuilder('goods')
      .leftJoinAndSelect('goods.producer', 'producer')
      .orderBy('goods.name', 'ASC')
      .skip(offset)
      .take(limit)
      .getManyAndCount();
  }

  const totalPages = Math.ceil(totalCount / limit);

  const response: GoodsResponse[] = goods.map(goodsItem => ({
    ID: goodsItem.id,
    Name: goodsItem.name,
    Image: goodsItem.image,
    IsInStock: goodsItem.isInStock,
    Description: goodsItem.description,
    Price: goodsItem.price,
    Producer: null,
    Tags: null,
    IsPrescriptionNeeded: null,
    Instruction: null
  }));

  return {
    Items: response,
    TotalPages: totalPages,
    CurrentPage: page,
  };
}

export async function getGoodsByID(
  dataSource: DataSource,
  id: number
): Promise<GoodsResponse> {
  const goodsRepository = dataSource.getRepository(Goods);
  
  const good = await goodsRepository.findOne({
    where: { id },
    relations: ['producer', 'goodsTags', 'goodsTags.tag'],
  });

  if (!good) {
    throw new Error('Good not found');
  }

  const tagNames = good.goodsTags?.map(goodsTag => goodsTag.tag?.tagName).filter(Boolean) || [];

  const response: GoodsResponse = {
    ID: good.id,
    Name: good.name,
    Image: good.image,
    Producer: good.producer?.producerName || '',
    IsInStock: good.isInStock,
    Tags: tagNames,
    Instruction: good.instruction,
    Description: good.description,
    IsPrescriptionNeeded: good.isPrescriptionNeeded,
    Price: good.price,
  };

  return response;
}

export async function getPromoItems(dataSource: DataSource): Promise<PromoItem[]> {
  const goodsRepository = dataSource.getRepository(Goods);
  
  const goods = await goodsRepository.find({
    where: { isInStock: true },
  });

  if (goods.length === 0) {
    return [];
  }

  const shuffled = [...goods];
  for (let i = shuffled.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [shuffled[i], shuffled[j]] = [shuffled[j], shuffled[i]];
  }
  
  const promoCount = Math.min(5, shuffled.length);
  const promoItems: PromoItem[] = shuffled.slice(0, promoCount).map(item => ({
    id: item.id,
    name: item.name,
    description: item.description,
    price: item.price,
    image: item.image,
  }));

  return promoItems;
}

export async function updateGoods(
  dataSource: DataSource,
  id: number,
  updateData: GoodsUpdateRequest,
  claims: Claims
): Promise<string> {
  const goodsRepository = dataSource.getRepository(Goods);
  
  const existingGood = await goodsRepository.findOne({
    where: { id },
  });

  if (!existingGood) {
    console.log(`Good error [${claims.username}] Good not found with id: ${id}`);
    throw new Error('Good not found');
  }

  existingGood.name = updateData.Name;
  existingGood.instruction = updateData.Instruction;
  existingGood.description = updateData.Description;
  existingGood.isPrescriptionNeeded = updateData.IsPrescriptionNeeded;
  existingGood.isInStock = updateData.IsInStock;
  existingGood.price = updateData.Price;

  await goodsRepository.save(existingGood);

  console.log(`Good PATH [${claims.username}] - Updated good ID: ${id}`);

  return "Good updated";
}